File: Python 实现/test_05.py
"""
题目描述
用两个栈来实现一个队列，完成队列的Push和Pop操作。 队列中的元素为int类型。
"""
"""
解题思路
第一个栈stack_1临时保存插入的数据，当调用弹出函数时候，若stack_2不为空则直接弹出，
若为空，则把stack_1中的数据全部弹出放到stack_2中。
注：stack_2都是存放的旧的数据，弹出时一定符合队列的规律。
"""
class Solution:
    def __init__(self):
        self.stack_1 = []
        self.stack_2 = []

    def push(self,node):
        self.stack_1.append(node)

    def pop(self):
        if self.stack_2:
            return self.stack_2.pop()
        else:
            while self.stack_1:
                self.stack_2.append(self.stack_1.pop())
            return self.stack_2.pop()

if __name__ == '__main__':
    s = Solution()
    s.push(1)
    s.push(2)
    print(s.stack_1)
    s.pop()
    print(s.stack_2)
    print(s.stack_1)

File: C++ 实现/Test_05.cpp
#include<iostream>
#include<stack>
using namespace std;
 
class Solution
{
public:
	//先将数依次进栈 
    void push(int node) {
        stack1.push(node);
    }
    int pop() {
    	//再将栈stack1中的数依次弹入到栈 stack2中 
        while(!stack1.empty())
        {
            stack2.push(stack1.top());
            stack1.pop();
        }
        //返回栈2栈顶数给a，做pop()的返回值 
        int a=stack2.top();
        stack2.pop();
        //再将栈2的数依次弹入到栈1中 
         while(!stack2.empty())
        {
            stack1.push(stack2.top());
            stack2.pop();
        }   
        return a;
    }
private:
    stack<int> stack1;
    stack<int> stack2;
};

int main()
{
	int n;
 	Solution s;
 	//将6个int数进栈 
 	for(int i=0;i<6;i++)
 	{
 		cin>>n;
 		s.push(n);
 	}
 	//调用pop()函数模拟队列先进先出 
 	for(int i=0;i<6;i++)
	 {
	 	cout<<s.pop()<<endl;
	 }		 
 	return 0;
}
